README.md
# pypdns

pypdns is a simple library wrapper around PDNS api (https://doc.powerdns.com/md/httpapi/README/).
It also feature a cli application for interacting with PowerDNS API.

  - Configurable via .ini or cli arguments
  - depends on : requests, docopt
  - Still a work in progress...

pypdns is not an official PowerDNS tool.

### Installation

pypdns is a standard python package, currently not push on pypi. Feel free to use your preferred way of installing.
(packaging is coming)

### Configuration (cli)

You can provide a  configuration file (check pypdns.example.ini), pypdns will first look in the current directory for a file name ```pypdns.ini``` or you can provide a path to a conf file through cli argument ```--config```.

### Usage

#### Creation


```pypdns  zones create example.com. --soa 'ns1.example.com. admins.example.com. 0 28800 7200 604800 86400' --nameservers=ns1.example.com.,ns2.example.com. --kind MASTER```

Note the serial is 0 , it's because serial is managed by PowerDNS automatically.

the --kind parameter is by default : NATIVE , if you want your zone to be publicly exposed, you have to specify --kind=MASTER.


#### Edition

When editing a record, the API always receive a PATCH requests with a full body (re)-defining the record ( creates, edits, disabled, delete ), by defaults , we first check if the record exists before making any changes.

the --override options is used for controlling the behaviour :

if set to True : No check before making the requests → all records existing records will be replaced

if set to False ( default ) : In a CLI mode, we asks the users if he wants to continue and we display the record which going to be replaced, when using the library, you MUST specify override kwargs otherwise the function exist with err = -1.

Note that for all operations : Delete, disabled , you can specify --override if you know what you're doing since the record is obviously present in the database.

##### Add record

Short method:
``` pypdns record add test.example.com 10.0.0.0 'Add record test' A --reverse```

Long method:
```pypdns  record edit example.com test 10.190.26.57 'Add record test' --rtype A --reverse```

Use --reverse to automatically add a PTR record for the newly created record. Warning ! if a previous PTR record exists it will be replaced by the new one.
Disable record

```pypdns  record edit example.com test 10.190.26.57 'Disable record test' --rtype A --reverse --disabled```

This command disabled the given record , returning NXDOMAIN on resolve. But the record is still present in the DB.

##### Re-enable record

```pypdns  record edit example.com test 10.190.26.57 'Add record test' --rtype A ```

##### Delete record

Short method
``` pypdns record delete test.example.com 10.190.26.57 'Delete test record' A --override```

Long method:
```pypdns  record edit example.com test 10.190.26.57 'Delete record test' --rtype A --changetype DELETE --override```

Note that the PTR record ( if it exists ) will be not delete even if you specify --reverse (smile)

##### Delete PTR record

Short method:
``` pypdns record delete 57.26.190.in-addr.arpa. 'Delete reverse test record' PTR --override```

Long method:
```pypdns  record edit 190.10.in-addr.arpa. 57.26 test.example.com 'Delete reverse test record' --rtype PTR --changetype DELETE --override```

#### Read

Note that all the filtering is done on client side so the requests will responds with all the data available.

The parameter --name is used to construct a regular expression, you can use that to filter with a pattern.

##### List zones

```pypdns  zones list --name <zone_name>```

List all the zones on the server, use --name to get a specific name.

##### Get zones

```pypdns  zones get <zone_name> --name <record_name> --rtype <record_type>```

Get zone content , you can filter on either record name or record type.

##### Search in Zone / Record

```pypdns  search <search_term>```

The * character can be used in <search_term> as a wildcard character and the ? character can be used as a wildcard for a single character.


debian/changelog
pypdns (0.17-4) unstable; urgency=medium

  * New upstream release
  * Fix broken python3 packaging

 -- Ismael Tifous <ismael.tifous@gmail.com>  Mon, 09 Oct 2017 15:59:31 +0200

pypdns (0.16-3) unstable; urgency=medium

  * New upstream release
  * Add support for python3 packaging

 -- Ismael Tifous <ismael.tifous@gmail.com>  Fri, 06 Oct 2017 16:53:51 +0200

pypdns (0.15-2) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Thu, 21 Sep 2017 11:04:48 +0200

pypdns (0.14-1) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Wed, 20 Sep 2017 16:39:57 +0200

pypdns (0.13-1) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Mon, 18 Sep 2017 16:34:23 +0200

pypdns (0.12-3) unstable; urgency=medium

  * New upstream version

 -- Ismael Tifous <ismael.tifous@gmail.com>  Wed, 06 Sep 2017 16:59:04 +0200

pypdns (0.11-2) unstable; urgency=medium

  * New upstream version

 -- Ismael Tifous <ismael.tifous@gmail.com>  Fri, 01 Sep 2017 16:31:36 +0200

pypdns (0.10-1) unstable; urgency=medium

  * New upstream version

 -- Ismael Tifous <ismael.tifous@gmail.com>  Thu, 31 Aug 2017 17:36:31 +0200

pypdns (0.9-1) unstable; urgency=medium

  * New upstream

 -- Ismael Tifous <ismael.tifous@gmail.com>  Tue, 08 Aug 2017 09:07:41 +0200

pypdns (0.8-1) unstable; urgency=medium

  * New upstream

 -- Ismael Tifous <ismael.tifous@gmail.com>  Fri, 04 Aug 2017 17:50:15 +0200

pypdns (0.7-1) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Fri, 04 Aug 2017 17:20:05 +0200

pypdns (0.6-1) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Wed, 02 Aug 2017 09:23:00 +0200

pypdns (0.5-1) unstable; urgency=medium

  * New upstream release

 -- Ismael Tifous <ismael.tifous@gmail.com>  Thu, 29 Jun 2017 14:53:15 +0200

pypdns (0.4-1) unstable; urgency=medium

  * New upstream version

 -- Ismael Tifous <ismael.tifous@gmail.com>  Mon, 26 Jun 2017 15:21:08 +0200

pypdns (0.1-1) unstable; urgency=low

  * Initial package definition

 -- Ismael Tifous <ismael.tifous@gmail.com>  Fri, 02 Jun 2017 17:30:28 +0200


tests/pypdns_tests.py
import os

import pytest

from pypdns import PyPDNS


@pytest.fixture
def conf_from_file():
    path =  os.path.join(os.path.dirname(__file__), '..', 'pypdns-test.ini')
    return {'config_path': path}


@pytest.fixture
def conf_from_dict():
    return {
        'endpoint': 'http://pdns.test:8081',
        'apikey': 'testapikey'
    }


def test_config_from_file(conf_from_file):
    pdns_api = PyPDNS(conf_from_file)
    cfg = pdns_api.config
    assert cfg['endpoint'], 'http://localhost:8080'
    assert cfg['apikey'], 'myapikey'



pypdns/api.py
import requests
import logging


log = logging.getLogger(__name__)


class PdnsAPI(object):

    collection_url = ''
    url = ''
    _session = None

    def __init__(self, config):
        self.endpoint = config['endpoint']
        self.apikey = config['apikey']

    @property
    def session(self):
        if not self._session:
            self._session = requests.Session()
            self._session.headers.update({'X-Api-Key': self.apikey})
        return self._session

    def _process_resp(self, resp, sts_code):
        data = None
        try:
            data = resp.json()
        except ValueError:
            log.debug("No json response found.")
            data = resp.text

        return data, sts_code

    def build_url(self, _id, url):
        if not url and _id:
            url = self.endpoint + self.url.format(_id)
        elif not _id and not url:
            raise ValueError('Provide either an resource id or a full url to '
                             'call')
        return url

    def _call(self, verb, url, data={}, params={}, headers={}):
        if headers:
            self.session.headers.update(headers)

        log.info('Call %s with data %s', url, data)
        resp = self.session.request(verb, url, json=data, params=params)
        status_code = resp.status_code
        try:
            resp.raise_for_status()
        except requests.HTTPError as exc:
            log.debug('Error occured, endpoint : %s, apikey : %s',
                      url, self.apikey)
            return resp, status_code

        except requests.Timeout:
            log.error('Request Timeout to %si', url)
            return False, status_code
        except requests.RequestException:
            log.error('Requests Error')
            return False, status_code
        else:
            return resp, status_code

    def collection_get(self, params={}):
        url = self.endpoint + self.collection_url
        return self._call('GET', url, params=params)

    def get(self, url=None, _id=None, params={}):
        url = self.build_url(_id, url)
        return self._call('GET', url, params=params)

    def collection_post(self, data):
        url = self.endpoint + self.collection_url
        return self._call('POST', url, data=data)

    def patch(self, data, _id=None, url=None):
        url = self.build_url(_id, url)
        return self._call('PATCH', url, data=data)


class ZonesAPI(PdnsAPI):
    collection_url = '/zones'
    url = '/zones/{}'

    def get_zones(self):
        return self._process_resp(*self.collection_get())

    def get_zone(self, zone_id):
        return self._process_resp(*self.get(_id=zone_id))

    def create_zone(self, data):
        return self._process_resp(*self.collection_post(data))

    def update_records(self, zone_name, data):
        return self._process_resp(*self.patch(data, _id=zone_name))


class SearchAPI(PdnsAPI):
    collection_url = '/search-data'
    url = ''

    def search(self, term, max_results=None):
        params = {'q': term}
        if max_results:
            params['max'] = max_results
        return self._process_resp(*self.collection_get(params=params))


pypdns/pypdns.py
import os
import re
import logging
import datetime
from six.moves import configparser as ConfigParser
from six import iteritems
from builtins import range, input

from . import api


log = logging.getLogger(__name__)


DEFAULT_CONFIG = {
    'endpoint': 'http://localhost:8081/api/v1/servers/localhost',
}


def load_config(custom_path=''):
    config = ConfigParser.ConfigParser()
    paths = ['./pypdns.ini',
             '~/.config/pypdns.ini']
    if custom_path:
        paths.insert(0, os.path.realpath(custom_path))
    config.read(paths)
    return config


def parse_config(cfg, ext_config={}):
    config = {}
    try:
        options = cfg.options('pypdns')
    except ConfigParser.NoSectionError:
        log.debug('No config files provided and default not present')
        options = []

    for option in options:
        config[option] = cfg.get('pypdns', option)

    for ckey, cvalue in iteritems(DEFAULT_CONFIG):
        if ckey not in config:
            config[ckey] = cvalue

    for k in ('endpoint', 'apikey'):
        if ext_config.get(k) is not None:
            config[k] = ext_config[k]

    assert (config.get('endpoint') is not None and
            config.get('apikey') is not None), 'Configuration not found'

    return config


class PyPDNS(object):

    _interactive = False

    def __init__(self, ext_config={}):
        cfg = load_config(ext_config.get('config_path'))

        self.config = parse_config(cfg, ext_config)
        self.zones_api = api.ZonesAPI(self.config)
        self.search_api = api.SearchAPI(self.config)

    def zones_list(self, name='.*'):
        """
        Get all zone from remote pdns API

        :return: List of zone
        :rtype:
        """
        ret, code = self.zones_api.get_zones()
        if code != 200:
            return ret

        result = []
        name = re.compile(name)
        for zone in ret:
            match = name.match(zone['name'])
            if match:
                result.append(zone)
        return result

    def zones_get(self, zone_id, name='.*', _type='.*'):
        """
        Return zone data.

        :param zone_id: Zone name
        :type zone_id: String
        :param name: Filter rrsets of given zone on name
        :type name: String
        :param _type: Filter rrsets of given zone on type
        :type _type: String
        :return: zone data as dict or False if not found.
        :rtype: dict
        """
        ret, code = self.zones_api.get_zone(zone_id)
        patterns = {}
        result = []
        if code != 200:
            return ret
        patterns['name'] = re.compile(name)

        patterns['type'] = re.compile(_type)

        for record_set in ret['rrsets']:
            match = True
            for key, pattern in iteritems(patterns):
                match &= bool(pattern.match(record_set[key]))

            for comment in record_set.get('comments', []):
                if comment['modified_at']:
                    comment['modified_at'] = datetime.datetime.fromtimestamp(
                        comment['modified_at']).isoformat()
            if match:
                result.append(record_set)

        return result

    def zones_create(self, name, kind='NATIVE', soa=None, nameservers=[],
                     soa_edit='INCEPTION-INCREMENT', soa_ttl=7200):
        """
        Creates a Zone

        :param name: Zone name
        :type name: String
        :param soa: Start of authority, override default config from pdns server
        :type name: String
        :param nameservers: List of nameservers for the given zone
        :type nameservers: list
        :param soa_edit: SOA edit behaviour for serial update
        :type soa_edit: String
        """
        soa = soa or self.config.get('default-soa')
        nameservers = nameservers or self.config.get('nameservers')
        if not (soa and nameservers):
            raise ValueError('Missing soa or nameservers value')

        data = {
            'name': name,
            'kind': kind,
            'nameservers': nameservers,
            'soa_edit_api': soa_edit,
            'rrsets': [
                {'name': name,
                 'type': 'SOA',
                 'ttl': soa_ttl,
                 'records': [{'content': soa, 'disabled': False}]}
            ]
        }
        ret, code = self.zones_api.create_zone(data)
        return ret, code

    def record_add(self, zone_name, record_name, contents, comment, type_='A',
                   changetype='REPLACE', ttl=3600, reverse=False, disabled=False,
                   override=False):
        """
        Add or replace a record, Return 2-tuple : response body, response code
        if record exists and override is set to False : return error message
        and -1 error code.

        :param zone_name: Zone to add record to
        :type zone_name: String
        :param record_name: Name of the record
        :type record_name: String
        :param content: Content of the record
        :type content: String
        :param type_: record's type
        :type type_: String
        :param changetype: Update behaviour, REPLACE or DELETE, default to REPLACE
        :type changetype: String
        :param ttl: record's type, default to A
        :type ttl: int
        :param reverse: For A or AAAA record, default is to False, set this to True for creating a reverse record
        :type reverse: bool
        :param disabled: Disable record, default to False
        :type disabled: bool
        """
        record_name = (record_name if record_name.endswith('.')
                       else record_name + '.')

        type_ = type_.upper()

        name = (record_name + zone_name if (type_ not in ('SOA',) and
                                            record_name != '.')
                else zone_name)

        if not override:
            # record already exists ?
            existing_record = self._get_record(name, type_)
            if existing_record:
                if not self._interactive:
                    return ('Specify override=True for erasing existing record',
                            -1)
                else:
                    valid = self._validate_override(existing_record)
                    if not valid:
                        return False

        if not name.endswith('.'):
            name += '.'

        records = []

        for content in contents.split(','):
            record = {
                'content':  content,
                'set-ptr':  reverse,
                'disabled': disabled
            }
            records.append(record)

        data = {
            'rrsets': [
                {'name': name,
                 'type': type_,
                 'ttl': ttl,
                 'changetype': changetype,
                 'records': records,
                 'comments': [{'account': os.environ['USER'],
                               'content': comment,
                               }
                              ]
                 }
            ]
        }
        ret, code = self.zones_api.update_records(zone_name, data)
        return ret, code

    def search(self, term, object_type=None, zone=None, rtype=None,
               max_results=None):
        """
        Search for data in zone & record

        :param term: term to search, use * for wildcard char and ? for a single wildcard char
        :type term: String
        :param object_type: filter results, zone or record
        :type object_type: String
        :param zone: filter results record on a specific zone (sets object_type = record)
        :type zone: String
        :param rtype: filter on the given record's type (implies object_type = record)
        :param max_results: Limit number of results, default to None
        :rtype max_results: int
        """
        results, sts_code = self.search_api.search(term, max_results)
        object_type = 'record' if zone or rtype else object_type
        if not object_type or sts_code != 200:
            return results

        filtered = [rec for rec in results if rec['object_type'] == object_type]

        if zone:
            zone = zone + '.' if not zone.endswith('.') else zone
            filtered = [rec for rec in filtered if rec.get('zone') == zone]

        if rtype:
            filtered = [rec for rec in filtered if rec.get('type') == rtype]

        return filtered

    def add(self, record_name, content, comment, rtype,ttl=3600, reverse=False,
            override=False):
        record, zone = self._construct_names(record_name)
        if not zone:
            log.info('Cannot find zone for record : %s', record_name)
            return False

        return self.record_add(zone, record, content, comment, type_=rtype,
                               ttl=ttl, reverse=reverse, override=override)

    def delete(self, record_name, comment, rtype, override=False):
        record, zone = self._construct_names(record_name)
        if not zone:
            log.info('Cannot find zone for record : %s', record_name)
            return False

        return self.record_add(zone, record, '', comment, type_=rtype,
                               override=override, changetype='DELETE')

    def _get_record(self, name, rtype):
        if name.endswith('.'):
            sname = name[:-1]
        else:
            sname = name
            name += '.'
        search = self.search(sname)
        return next((rr for rr in search if rr.get('name') == name and
                                            rr.get('type') == rtype),
                    None)

    def _validate_override(self, record):
        valid = False
        while not valid:
            replace = input('Warning, the record already exists with the '
                            'following data %s, replace it ? Y/n :'
                            '' % record)
            if replace.lower() not in ('y','n','yes','no'):
                continue
            if replace.lower() in ('n', 'no'):
                log.info('Exiting with no changes on existing record')
                valid = False
                break
            else:
                valid = True
        return valid

    def _zone_exists(self, zone):
        resp = self.zones_get(zone)
        if 'error' in resp:
            res = None
        else:
            res = resp
        return res

    def _construct_names(self, name):
        sub_names = name.split('.')
        # Iterates on sub domains to find record name & zone, use api to find
        # the first name matching a zone.
        zone = []
        record = []
        for i in range(0, len(sub_names)):
            sub_zone = '.'.join(sub_names[i:])
            search = self._zone_exists(sub_zone)
            if search:
                zone = sub_names[i:]
                record = [rr for rr in sub_names if rr not in sub_names[i:]]
                break
        return ('.'.join(record), '.'.join(zone))


pypdns/cli.py
"""
pypdns

Usage:
    pypdns zones list [--name <name>] [-c <cfg_pth>] [--log <log_level>]
    pypdns zones get <zone_name> [--name <name>] [--type <type>] [--log <log_level>] [-c <cfg_pth>]
    pypdns zones create <zone_name> [--soa <soa>] [--kind <kind>] [--nameservers=<nameservers>] [--soa-edit <soa_edit>] [--soa-ttl <soa_ttl>]
    pypdns record add <full_record_name> <content> <comment> <rtype> [--ttl <ttl>] [--reverse] [--disabled] [--override] [-c <cfg_pth>] [--log <log_level>]
    pypdns record delete <full_record_name> <comment> <rtype> [--override] [-c <cfg_pth>] [--log <log_level>]
    pypdns record edit <zone_name> <record_name> <content> <comment> --rtype <type> [--changetype <changetype>] [--ttl <ttl>] [--reverse] [--disabled] [--override] [-c <cfg_pth>] [--log <log_level>]
    pypdns search <term> [--otype <object_type>] [--zone <zone>] [--rtype <type>] [--max-results <max_results>] [-c <cfg_pth>] [--log <log_level>]
    pypdns (-h | --help)
    pypdns --version

Options:
    -c <cfg_pth> --config=<cfg_pth>  Path of configuration file, if not provided default to: ./pypdns.ini, ~/.config/pypdns.ini
    -h --help                        Show this screen
    --version                        Show version
    --endpoint <api_endpoint>        Override config file for pdns api endpoint
    --apikey <api_key>               Override config file for pdns api api key
    --log <log_level>                Set log level [default: ERROR]
    --name <name>                    Filter results on name for a given zone, regex format [default: .*]
    --type <type>                    Filter results on type for a given zone, regex format [default: .*]
    --soa <soa>                      SOA valid string, override default_soa in config file
    --soa-ttl <soa_ttl>              SOA ttl, [default: 7200]
    --kind <kind>                    Zone type, [default: NATIVE]
    --nameservers=<nameservers>      List of namservers for zone, override value from config file
    --soa_edit <soa_edit>            Soa edit behaviour for the zone, [default: DEFAULT]
    --rtype <type>                   Record's type
    --reverse                        Add reverse record for A and AAAA [default: False]
    --changetype <changetype>        Record update behaviour [default: REPLACE]
    --disabled                       Disable record [default: False]
    --override                       Override existing record, if no, a confirmation will be ask on prompt [default: False]
    --ttl <ttl>                      Record's ttl [default: 3600]
    --otype <object_type>            Filter search result, one of : record or zone [default: record]
    --zone <zone>                    Filter search result on zone name (sets object_type to record)
    --max-results <max_results>      Limit search results
"""
import logging
import json

from docopt import docopt

from . import __version__ as VERSION
from .pypdns import PyPDNS


def main():
    options = docopt(__doc__, version=VERSION)
    cfg = {
        'config_path': options.get('--config'),
        'endpoint': options.get('--endpoint'),
        'apikey': options.get('--apikey')
    }
    logging.basicConfig()
    log_level = getattr(logging, options['--log'].upper(), None)
    if not isinstance(log_level, int):
        raise ValueError('Invalid log level: %s' % log_level)

    logger = logging.getLogger()
    logger.setLevel(log_level)
    handler = logging.StreamHandler()
    formatter = logging.Formatter(
        '%(asctime)s %(name)-12s %(levelname)-8s %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.info('Start')
    pdns_api = PyPDNS(cfg)
    # We are in CLI mode
    pdns_api._interactive = True

    if 'zones' in options and options['zones']:
        if 'list' in options and options['list']:
            print(json.dumps(pdns_api.zones_list(options['--name']), sort_keys=True, indent=4, separators=(',', ': ')))

        if 'get' in options and options['get']:
            print(json.dumps(pdns_api.zones_get(options['<zone_name>'],
                                                options['--name'], options['--type']),
                             sort_keys=True, indent=4, separators=(',', ': ')))

        if 'create' in options and options['create']:
            nameservers = options.get('--nameservers', [])
            nameservers = nameservers and nameservers.split(',')

            print(json.dumps(pdns_api.zones_create(options['<zone_name>'],
                                         kind=options['--kind'],
                                         soa=options['--soa'],
                                         nameservers=nameservers,
                                         soa_edit=options['--soa-edit'],
                                         soa_ttl=options['--soa-ttl']),
                             sort_keys=True, indent=4, separators=(',', ': ')))

    if options['record']:
        if options['add']:
            print(json.dumps(pdns_api.add(options['<full_record_name>'],
                                          options['<content>'],
                                          options['<comment>'],
                                          options['<rtype>'],
                                          ttl=int(options['--ttl']),
                                          reverse=options['--reverse'],
                                          override=options['--override'])))

        if options['delete']:
            print(json.dumps(pdns_api.delete(options['<full_record_name>'],
                                             options['<comment>'],
                                             options['<rtype>'],
                                             options['--override'])))
        if options['edit']:
            print(json.dumps(pdns_api.record_add(options['<zone_name>'],
                                                 options['<record_name>'],
                                                 options['<content>'],
                                                 options['<comment>'],
                                                 type_=options['--rtype'],
                                                 changetype=options['--changetype'],
                                                 ttl=int(options['--ttl']),
                                                 reverse=options['--reverse'],
                                                 disabled=options['--disabled'],
                                                 override=options['--override']),
                             sort_keys=True, indent=4, separators=(',', ': ')))
    if options['search']:
        print(json.dumps(pdns_api.search(options['<term>'],
                                         object_type=options['--otype'],
                                         zone=options['--zone'],
                                         rtype=options['--rtype'],
                                         max_results=options['--max-results']),
                         sort_keys=True, indent=4, separators=(',', ': ')))
